analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion.RoslynCfg.cs

```diff
@@ -0,0 +1,179 @@
+                if (!walker.CheckPaths() && !cfg.ExitBlock.IsReachable)
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, identifier.GetLocation(), "method's recursion"));
+                }
+            }
+
+            private class CommonRecursionSearcher : CfgAllPathValidator
+            {
+                private readonly ISymbol memberToCheck;
+                private readonly Action reportIssue;
+                private readonly ControlFlowGraph cfg;
+                private readonly bool isGetAccesor;
+
+                public CommonRecursionSearcher(RecursionContext<ControlFlowGraph> context, bool isGet = true)
+                    : base(context.ControlFlowGraph)
+                {
+                    memberToCheck = context.AnalyzedSymbol;
+                    reportIssue = context.ReportIssue;
+                    cfg = context.ControlFlowGraph;
+                    isGetAccesor = isGet;
+                }
+
+                public bool CheckPaths()
+                {
+                    if (CheckAllPaths())
+                    {
+                        reportIssue();
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                protected override bool IsValid(BasicBlock block) =>
+                    FindReferenceToSelf(block);
+
+                protected override bool IsInvalid(BasicBlock block) => false;
+
+                private bool FindReferenceToSelf(BasicBlock block)
+                {
+                    foreach (var operation in block.OperationsAndBranchValue)
+                    {
+                        if (ProcessOperation(operation, cfg, out var result))
+                        {
+                            return result;
+                        }
+                    }
+                    return false;
+
+                    bool ProcessOperation(IOperation operation, ControlFlowGraph controlFlowGraph, out bool result)
+                    {
+                        foreach (var child in operation.DescendantsAndSelf().ToReversedExecutionOrder())
+                        {
+                            if (child.Instance.Kind == OperationKindEx.FlowAnonymousFunction)
+                            {
+                                var anonymousFunctionCfg = controlFlowGraph.GetAnonymousFunctionControlFlowGraph(IFlowAnonymousFunctionOperationWrapper.FromOperation(child.Instance));
+                                foreach (var subOperation in anonymousFunctionCfg.Blocks.SelectMany(x => x.OperationsAndBranchValue).SelectMany(x => x.DescendantsAndSelf()))
+                                {
+                                    if (ProcessOperation(subOperation, anonymousFunctionCfg, out result))
+                                    {
+                                        return true;
+                                    }
+                                }
```

Comment:
_Thinking out loud:_ This part is very similar to `anonymousFunctionCfg.Blocks.Any(FindReferenceToSelf)`, it just returns a little different result. 

Can we unify them somehow? It's complicated so I'd like to have it just once there.

Reply:
Part was not needed and got removed